### ISSUE TITLE
100% CPU after upgraded to v1.35

### ISSUE THREAD

**Issue by clusc** [aerospike/aerospike-client-go#242]
We've noticed a blocking issue after migrate to v1.35. Our loadtest cpu util goes from 20% -> 100%, and our service becomes unusable at all. I've taken a pprof capture, it's all pointing in cipher.go. Server is 3.14.1.2, client go 1.10.3. Please help us investigate what's going on. 

![image](https://user-images.githubusercontent.com/25391235/45340698-77643d00-b54c-11e8-9f46-1fcc89abc5db.png)

**Comment by clusc** [comment]
[pprof.server.samples.cpu.001.pb.gz](https://github.com/aerospike/aerospike-client-go/files/2369472/pprof.server.samples.cpu.001.pb.gz)

**Comment by khaf** [comment]
Let me see if I can easily reproduce the issue. 
The cypher.go itself is an external library and hasn’t been changed in a long time.

**Comment by khaf** [comment]
@clusc it seems that a lot of connections are being dropped, and as a result new connections are being established. Each of those will require to authenticate again, annd what we see in this profile is a symptom of that.

Can you share the results of `client.Stats()` after a few minutes of your application running?

**Comment by clusc** [comment]
No we are using simple username/password authentication. I do see lots of timeouts while tailing the log, and it's like an avalanche effect where timeout drives the cpu utils up, and more timeouts are happening. Can you think of any reason why this would occurs since we upgrade to 1.35?

Is there a way to instrument `client.Stats()` dynamically? Or you suggest I added a timer and print it?

**Comment by khaf** [comment]
I think the easiest way would be to print it via a simple timer.

The question here is why you are encountering so many timeouts. One possibility is that your timeouts are very tight, and the number of goroutines generated via batch commands is overwhelming the scheduler to the point that it cannot service a big percentage on time.

Is there anything else that has changed in your setup during this round of upgrades?

**Comment by clusc** [comment]
We do have some timeouts occasionally. What I see from eyeballing the log is that, it's been fine for the first a few sec, and as soon as we have some timeouts, it becomes bad and everything becomes unfunctional.

No we didn't change anything, except for upgrade the client lib version. I can verify it by switching back and forth of the client version, and the issue is 100% reproducible. Our loadtest runs at a constant rate, so this is very interesting. Is there anything change regarding client connection handling in the new version? The last version we were on is 0a56b2c463372510e837a064caa05f6c47a3c255.

**Comment by khaf** [comment]
Is it possible to post a simple gist that is close to your usage pattern of the API that reproduces the issue? My simple tests have been unsuccessful until now.

It is possible that your specific cluster setup and configuration are also relevant. Is it possible to share the server config?

**Comment by clusc** [comment]
Our call is purely batch_get with 3s connection timeout, 500ms base timeout, 250ms socket timeout, 3 retries. 

Traffic is about 30k batch_sub_read_success & 16k batch_sub_read_not_found.

I attached the server config for reference.
[aerospike.cfg.zip](https://github.com/aerospike/aerospike-client-go/files/2385181/aerospike.cfg.zip)

Not sure what's the best way to get to the bottom of this.

**Comment by clusc** [comment]
Attached the flame graph for both good and bad state. I initially think this has something to do with server version, so I went ahead to upgrade from 3.14 to 4.3.0.7, but the issue remains. This is blocking our feature deployment now, please help us investigate asap. Thanks

[aerospike.zip](https://github.com/aerospike/aerospike-client-go/files/2390825/aerospike.zip)

**Comment by khaf** [comment]
The timeouts threw me off a bit. The whole time I was looking for the type of complex problems we have encountered in the past.

Turned out this was a simple regression which only affected performance for enterprise users who utilize the authentication.

It was a regression in which bcrypt was used to hash the password for each login, which is very slow (exactly as the profile showed). Now it is fixed to use the hashed cache.

The fix is released.
